File: 2020/python/day01.py

def main():
  xs = read_input()
  print(f'part 1: {part_1(xs)}')
  print(f'part 2: {part_2(xs)}')

def part_1(xs):
  return [
    xs[i] * (2020 - xs[i])
    for i in range(len(xs)-1)
    if 2020 - xs[i] in xs
  ][0]

def part_2(xs):
  return [
    xs[i] * xs[j] * (2020 - xs[i] - xs[j])
    for i in range(len(xs)-1)
    for j in range(i+1, len(xs))
    if 2020 - xs[i] - xs[j] in xs
  ][0]

#####################################################################

def read_input():
  with open('../inputs/01.txt') as f:
    return [int(line) for line in f.readlines()]

#####################################################################

if __name__ == "__main__":
  main()

File: 2020/haskell/Day01.hs
module Day01 where

import Text.Printf(printf)

main = do
  input <- readFile "../inputs/01.txt"
  printf "part 1: %d\n" $ part1 $ linesToInts input
  printf "part 2: %d\n" $ part2 $ linesToInts input

part1 xs = [
    xs!!i * (2020 - xs!!i) | 
    i <- [0..(length xs - 1)], 
    elem (2020 - xs!!i) xs
  ]!!0

part2 xs = [
    xs!!i * xs!!j * (2020 - xs!!i - xs!!j)| 
    i <- [0..(length xs - 2)], 
    j <- [i+1..(length xs - 1)], 
    elem (2020 - xs!!i - xs!!j) xs
  ]!!0

-- For both parts --
linesToInts :: String -> [Int]
linesToInts a = map read (lines a)
